fdbclient/VersionVector.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * VersionVector.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FDBCLIENT_VERSION_VECTOR_H
+#define FDBCLIENT_VERSION_VECTOR_H
+
+#pragma once
+
+#include <unordered_map>
+
+#include "fdbclient/FDBTypes.h"
+
+struct VersionVector {
+	std::unordered_map<Tag, Version> versions;
+	Version maxVersion; // Specifies the max version in this version vector. (Note:
+                        // there may or may not be a corresponding entry for this
+                        // version in the "versions" map.)
+
+
+	VersionVector() : maxVersion(invalidVersion) {}
+	VersionVector(Version version) : maxVersion(version) {}
+
+	void setVersion(const Tag& tag, Version version) {
+        ASSERT(tag != invalidTag);
+		ASSERT(version > maxVersion);
```

Comment:
I can think a scenario this assertion would fail: consider a version vector cache at proxy (or client), where this function is called to update a storage tag. The version is not necessarily large than the `maxVersion`, because it could belong to another storage server.

Reply:
The way I was thinking about this, only the sequencer should update the version vector. A proxy/client will update/sync its cache with the sequencer's version vector on the read path.

Reply:
This is fine for now. I think in the future you'll want send deltas between two vectors and at that time this function can be used for applying the deltas.

Reply:
Yes.